serialization.rst

```diff
@@ -0,0 +1,337 @@
+
+If an object is of a class that implements the ``MongoDB\BSON\Serializable``
+interface, call ``bsonSerialize`` and used the returned associative array to
+store as properties of a **BSON document**. Not returning an array is an error
+and should throw an ``MongoDB\Driver\Exception\UnexpectedValueException``
+exception. It is valid to return a packed array, but it must also be stored as
+a **BSON document**.
+
+If an object is of a class that implements the ``MongoDB\BSON\Persistable``
+interface (which implies ``MongoDB\BSON\Serializable``), obtain the properties
+in a similar way as in the previous paragraph, but *also* add an additional
+property ``__pclass`` as a Binary value, with subtype ``0x80`` and as value
+the fully qualified class name of the object that is being serialized.
+
+Examples
+~~~~~~~~
+
+::
+
+  stdClass {
+    public $foo => 42
+  } => { 'foo' : 42 }
+
+  MyClass {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+  } => { 'foo' : 42 }
+
+  AnotherClass implements Serializable {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+    function bsonSerialize() : array {
+        return [ 'foo' => $this->foo, 'prot' => $this->prot ];
+    }
+  } => { 'foo' : 42, 'prot' : "wine" }
+
+  UpperClass implements Persistable {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+    function bsonSerialize() : array {
+        return [ 'foo' => $this->foo, 'prot' => $this->prot ];
+    }
+  } => { 'foo' : 42, 'prot' : "wine", '__pclass' : Binary(0x80, "UpperClass") }
+
+
+Deserialization from BSON
+=========================
+
+For compound types, there are three data types:
+
+- ``root``: refers to the top-level BSON document *only*
+- ``document``: refers to nested BSON document *only*
+- ``array``: refers to a BSON array
+
+Each of those three data types can be mapped against different PHP types. The
+possible mapping values are:
+
+- *not set* or ``NULL`` — this is the default.
+
+  - A BSON array will be deserialized as a PHP ``array``.
+  - A BSON document (root or nested) without ``__pclass`` property [1]_ becomes a
+    PHP ``stdClass`` object, with each BSON document property becoming a
+    public ``stdClass`` property.
+  - A BSON document (root or nested) with ``__pclass`` property [1]_ becomes
+    a PHP object of the class name as defined by the ``__pclass`` property.
+
+    If the named class implements the ``MongoDB\BSON\Unserializable``
+    interface, then the properties of the BSON document, minus the
+    ``__pclass`` property are sent as an associative array to the
+    ``bsonUnserialize`` function to initialise the object's properties.
+    
+    If the named class does not implement the ``MongoDB\BSON\Unserializable``
+    interface, then the object gets initialized with the properties
+    **without** the ``__pclass`` property being one of them. It simply gets
+    left out.
```

Comment:
We can't.. What if its an internal class for example?
It *must* implement the interface or we open ourself to can of problems.